File: packages/sdk-node/test/conformance.test.ts
import { generateKeyPairSync } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import { RateLimiter } from '../src/core/rate-limiter.js';
import { ShardedLimiter } from '../src/core/sharded-limiter.js';
import { privateKeyFromRaw, signingPayload } from '../src/core/budget-attestation.js';
import { EvidenceChain, GENESIS_PREV_HASH } from '../src/core/evidence-chain.js';
import { issueSpendReceipt } from '../src/core/spend-receipt.js';
import { extractTokenUsageFromText } from '../src/core/utils.js';

interface ConformanceVectors {
  policy: { requests_per_second: number; burst: number };
  steps: Array<{ note: string; advance_ms: number; n: number; allowed: boolean; remaining: number; retry_after_ms: number }>;
}

interface ExpiryVectors {
  cases: Array<{ note: string; epoch_ms: number; expected: string }>;
}

const vectorsPath = fileURLToPath(new URL('../../../conformance/token_bucket_vectors.json', import.meta.url));
const vectors: ConformanceVectors = JSON.parse(readFileSync(vectorsPath, 'utf-8'));

const expiryVectorsPath = fileURLToPath(
  new URL('../../../conformance/budget_attestation_expiry_vectors.json', import.meta.url),
);
const expiryVectors: ExpiryVectors = JSON.parse(readFileSync(expiryVectorsPath, 'utf-8'));

function fakeClock(startMs: number) {
  let now = startMs;
  return {
    clock: { now: () => now },
    advance: (ms: number) => {
      now += ms;
    },
  };
}

/**
 * Replays the same admission sequence used by the Go and Python SDKs against
 * the shared oracle in conformance/token_bucket_vectors.json. A failure here
 * means Node has drifted from the documented cross-language behavior — not
 * just from its own past test suite.
 */
describe('Conformance: token bucket', () => {
  it('matches the shared oracle', () => {
    const { clock, advance } = fakeClock(0);
    const limiter = new RateLimiter({ clock, capacity: 1_000 });
    const options = {
      requestsPerSecond: vectors.policy.requests_per_second,
      burst: vectors.policy.burst,
      windowMs: 60_000,
      remoteRateLimitEndpoint: '',
      apiKey: undefined as string | undefined,
    };

    for (const [i, step] of vectors.steps.entries()) {
      advance(step.advance_ms);
      const d = limiter.increment('conformance-key', options, step.n);
      expect(d.allowed, `step ${i} (${step.note})`).toBe(step.allowed);
      if (step.allowed) {
        expect(d.remaining, `step ${i} (${step.note})`).toBe(step.remaining);
      } else {
        expect(d.retryAfterMs, `step ${i} (${step.note})`).toBe(step.retry_after_ms);
      }
    }
  });

  /**
   * ShardedLimiter is a different internal representation (a single
   * "bucket-full-at" timestamp instead of {tokens, last}) of the SAME
   * algorithm. Replaying the shared oracle against it too is exactly what
   * proves decision parity, not just "looks similar."
   */
  it('matches the shared oracle via ShardedLimiter', () => {
    const { clock, advance } = fakeClock(0);
    const limiter = new ShardedLimiter({ clock });
    const options = {
      requestsPerSecond: vectors.policy.requests_per_second,
      burst: vectors.policy.burst,
      windowMs: 60_000,
      remoteRateLimitEndpoint: '',
      apiKey: undefined as string | undefined,
    };

    for (const [i, step] of vectors.steps.entries()) {
      advance(step.advance_ms);
      const d = limiter.increment('conformance-key', options, step.n);
      expect(d.allowed, `step ${i} (${step.note})`).toBe(step.allowed);
      if (step.allowed) {
        expect(d.remaining, `step ${i} (${step.note})`).toBe(step.remaining);
      } else {
        expect(d.retryAfterMs, `step ${i} (${step.note})`).toBe(step.retry_after_ms);
      }
    }
  });
});

/**
 * Replays the shared oracle in conformance/budget_attestation_expiry_vectors.json
 * against signingPayload, proving Node formats expires_at identically to Go
 * and Python inside the Ed25519 signing payload — not just that Node's own
 * round-trip tests pass. A failure here means a cross-language attested
 * budget token would fail to verify.
 */
describe('Conformance: budget attestation expiry formatting', () => {
  it('matches the shared oracle', () => {
    const { publicKey } = generateKeyPairSync('ed25519');

    for (const [i, tc] of expiryVectors.cases.entries()) {
      const grant = {
        maxTokens: 100,
        maxDepth: 1,
        expiresAt: new Date(tc.epoch_ms),
      };
      const raw = signingPayload(grant, publicKey);
      const decoded = JSON.parse(raw.toString('utf8'));
      expect(decoded.expires_at, `case ${i} (${tc.note})`).toBe(tc.expected);
    }
  });
});

// ── Evidence chain ──
//
// Rule 13: parity claims must be conformance-tested, not assumed. The entry
// hash covers a compact-JSON payload; Go, Node, and Python must produce the
// same bytes, and therefore the same hashes and head, from the same inputs.
// The vectors were generated by the Go reference implementation.

interface EvidenceChainVectors {
  seed_hex: string;
  issued_at_unix: number;
  genesis_prev_hash: string;
  claims: Array<{
    key: string;
    provider: string;
    model: string;
    window_start_unix: number;
    window_end_unix: number;
    input_tokens: number;
    output_tokens: number;
    total_tokens: number;
    estimated_cost_micro_usd: number;
  }>;
  entries: Array<{
    seq: number;
    prev_hash: string;
    receipt_signature_b64: string;
    entry_hash: string;
  }>;
  chain_head: string;
  total_tokens: number;
  total_estimated_cost_micro_usd: number;
}

describe('Conformance: evidence chain', () => {
  it('reproduces the Go reference entry hashes and chain head', () => {
    const path = fileURLToPath(
      new URL('../../../conformance/evidence_chain_vectors.json', import.meta.url),
    );
    const v: EvidenceChainVectors = JSON.parse(readFileSync(path, 'utf-8'));
    expect(GENESIS_PREV_HASH).toBe(v.genesis_prev_hash);

    const priv = privateKeyFromRaw(Buffer.from(v.seed_hex, 'hex'));
    const chain = new EvidenceChain();

    v.claims.forEach((c, i) => {
      const receipt = issueSpendReceipt(
        priv,
        {
          key: c.key,
          provider: c.provider,
          model: c.model,
          windowStartUnix: c.window_start_unix,
          windowEndUnix: c.window_end_unix,
          inputTokens: c.input_tokens,
          outputTokens: c.output_tokens,
          totalTokens: c.total_tokens,
          estimatedCostMicroUSD: c.estimated_cost_micro_usd,
        },
        v.issued_at_unix,
      );
      const entry = chain.append(receipt);
      const want = v.entries[i]!;

      expect(entry.seq, `entry ${i} seq`).toBe(want.seq);
      expect(entry.prevHash, `entry ${i} prev_hash`).toBe(want.prev_hash);
      expect(entry.receipt.signature.toString('base64'), `entry ${i} signature`).toBe(
        want.receipt_signature_b64,
      );
      expect(entry.entryHash, `entry ${i} hash`).toBe(want.entry_hash);
    });

    expect(chain.head).toBe(v.chain_head);

    const pkg = chain.exportEvidence(v.issued_at_unix);
    expect(pkg.totalTokens).toBe(v.total_tokens);
    expect(pkg.totalEstimatedCostMicroUSD).toBe(v.total_estimated_cost_micro_usd);
  });
});

// ── Streaming usage extraction ──
//
// Rule 13: parity claims must be conformance-tested, not assumed. These
// vectors exist because Node and Python silently reported NO usage for the
// single-usage-event case (the OpenAI-compatible shape) while Go handled it
// correctly — a real divergence that every per-language suite passed through.

interface SSEUsageVectors {
  cases: Array<{
    name: string;
    note: string;
    sse: string;
    expect_found: boolean;
    expect_input_tokens: number;
    expect_output_tokens: number;
    expect_total_tokens: number;
  }>;
}

describe('Conformance: streaming usage extraction', () => {
  const sseVectors: SSEUsageVectors = JSON.parse(
    readFileSync(
      fileURLToPath(new URL('../../../conformance/sse_usage_vectors.json', import.meta.url)),
      'utf-8',
    ),
  );

  for (const c of sseVectors.cases) {
    it(c.name, () => {
      const usage = extractTokenUsageFromText(c.sse);

      if (!c.expect_found) {
        expect(usage === undefined || usage.totalTokens === 0, c.note).toBe(true);
        return;
      }

      expect(usage, `no usage extracted — ${c.note}`).toBeDefined();
      expect(usage!.inputTokens, c.note).toBe(c.expect_input_tokens);
      expect(usage!.outputTokens, c.note).toBe(c.expect_output_tokens);
      expect(usage!.totalTokens, c.note).toBe(c.expect_total_tokens);
    });
  }
});

File: packages/sdk-go/conformance_test.go
package rateguard

import (
	"context"
	"crypto/ed25519"
	"crypto/rand"
	"encoding/json"
	"os"
	"testing"
	"time"
)

type conformanceVectors struct {
	Policy struct {
		RequestsPerSecond int `json:"requests_per_second"`
		Burst             int `json:"burst"`
	} `json:"policy"`
	Steps []struct {
		Note         string  `json:"note"`
		AdvanceMs    int64   `json:"advance_ms"`
		N            float64 `json:"n"`
		Allowed      bool    `json:"allowed"`
		Remaining    int     `json:"remaining"`
		RetryAfterMs int64   `json:"retry_after_ms"`
	} `json:"steps"`
}

func loadConformanceVectors(t *testing.T) conformanceVectors {
	t.Helper()
	data, err := os.ReadFile("../../conformance/token_bucket_vectors.json")
	if err != nil {
		t.Fatalf("read shared conformance vectors: %v", err)
	}
	var v conformanceVectors
	if err := json.Unmarshal(data, &v); err != nil {
		t.Fatalf("parse shared conformance vectors: %v", err)
	}
	return v
}

// TestConformanceTokenBucket replays the same admission sequence used by the
// Node and Python SDKs against the shared oracle in
// conformance/token_bucket_vectors.json, for every in-process Store
// implementation. A failure here means Go has drifted from the documented
// cross-language behavior — not just from its own past test suite.
func TestConformanceTokenBucket(t *testing.T) {
	vectors := loadConformanceVectors(t)
	policy := PolicyPreset{
		RequestsPerSecond: vectors.Policy.RequestsPerSecond,
		Burst:             vectors.Policy.Burst,
	}
	ctx := context.Background()

	type fixture struct {
		name  string
		clock *fakeLimiterClock
		l     Store
	}
	newClock := func() *fakeLimiterClock {
		return &fakeLimiterClock{now: time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)}
	}
	memClock, shardClock := newClock(), newClock()
	stores := []fixture{
		{"memory", memClock, newMemoryLimiterWithClock(memClock, 1000)},
		{"sharded", shardClock, newShardedLimiterWithClock(shardClock, 1000)},
	}

	for _, tc := range stores {
		t.Run(tc.name, func(t *testing.T) {
			for i, step := range vectors.Steps {
				tc.clock.advance(time.Duration(step.AdvanceMs) * time.Millisecond)
				d, err := tc.l.Increment(ctx, "conformance-key", policy, step.N)
				if err != nil {
					t.Fatalf("step %d (%s): %v", i, step.Note, err)
				}
				if d.Allowed != step.Allowed {
					t.Fatalf("step %d (%s): allowed = %v, want %v", i, step.Note, d.Allowed, step.Allowed)
				}
				if d.Allowed && d.Remaining != step.Remaining {
					t.Fatalf("step %d (%s): remaining = %d, want %d", i, step.Note, d.Remaining, step.Remaining)
				}
				if !d.Allowed && d.RetryAfter.Milliseconds() != step.RetryAfterMs {
					t.Fatalf("step %d (%s): retry_after_ms = %d, want %d", i, step.Note, d.RetryAfter.Milliseconds(), step.RetryAfterMs)
				}
			}
		})
	}
}

// TestConformanceBudgetAttestationExpiry replays the shared oracle in
// conformance/budget_attestation_expiry_vectors.json against signingPayload,
// proving Go formats expires_at identically to Node and Python inside the
// Ed25519 signing payload — not just that each SDK's own round-trip tests
// pass. A failure here means a cross-language attested budget token would
// fail to verify.
func TestConformanceBudgetAttestationExpiry(t *testing.T) {
	data, err := os.ReadFile("../../conformance/budget_attestation_expiry_vectors.json")
	if err != nil {
		t.Fatalf("read expiry vectors: %v", err)
	}
	var vectors struct {
		Cases []struct {
			Note     string `json:"note"`
			EpochMs  int64  `json:"epoch_ms"`
			Expected string `json:"expected"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("parse expiry vectors: %v", err)
	}

	delegatePub, _, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatalf("generate delegate key: %v", err)
	}

	for i, tc := range vectors.Cases {
		grant := BudgetGrant{MaxTokens: 100, MaxDepth: 1, ExpiresAt: time.UnixMilli(tc.EpochMs).UTC()}
		raw := signingPayload(grant, delegatePub)
		var decoded struct {
			ExpiresAt string `json:"expires_at"`
		}
		if err := json.Unmarshal(raw, &decoded); err != nil {
			t.Fatalf("case %d (%s): decode signing payload: %v", i, tc.Note, err)
		}
		if decoded.ExpiresAt != tc.Expected {
			t.Fatalf("case %d (%s): expires_at = %q, want %q", i, tc.Note, decoded.ExpiresAt, tc.Expected)
		}
	}
}
